mathics/builtin/atomic/symbols.py

```diff
@@ -59,36 +59,134 @@ def _get_usage_string(symbol, evaluation, is_long_form: bool):
+        usagetext = DocumentationEntry(docstr, title).text()
         usagetext = re.sub(r"\$([0-9a-zA-Z]*)\$", r"\1", usagetext)
-        return usagetext
-    return None
+    return usagetext
+
+
+def show_definitions(symbol: Symbol, evaluation: Evaluation) -> list:
+    """Return a list of lines describing the definition of `symbol`"""
+    lines = []
+
+    def print_rule(
+        rule: Rule,
+        up: bool = False,
+        lhs: Callable = lambda k: k,
+        rhs: Callable = lambda r: r,
+    ):
+        evaluation.check_stopped()
+        if isinstance(rule, Rule):
+            r = rhs(
+                rule.replace.replace_vars(
+                    {"System`Definition": Expression(SymbolHoldForm, SymbolDefinition)}
+                )
+            )
+            lines.append(
+                Expression(
+                    SymbolHoldForm,
+                    Expression(
+                        up and SymbolUpSet or SymbolSet, lhs(rule.pattern.expr), r
+                    ),
+                )
+            )
+
+    def print_rules(definition: Definition):
+        """
+        Print all the rules associated
+        to a definition object
+        """
+        for rule in definition.ownvalues:
+            print_rule(rule)
+        for rule in definition.downvalues:
+            print_rule(rule)
+        for rule in definition.subvalues:
+            print_rule(rule)
+        for rule in definition.upvalues:
+            print_rule(rule, up=True)
+        for rule in definition.nvalues:
+            print_rule(rule)
+        formats = sorted(definition.formatvalues.items())
+        for format, rules in formats:
+            for rule in rules:
+
+                def lhs(expr):
+                    return Expression(SymbolFormat, expr, Symbol(format))
+
+                def rhs(expr):
+                    if expr.has_form("Infix", None):
+                        expr = Expression(
+                            Expression(SymbolHoldForm, expr.head), *expr.elements
+                        )
+                    return Expression(SymbolInputForm, expr)
+
+                print_rule(rule, lhs=lhs, rhs=rhs)
+
+    name = symbol.get_name()
+    if not name:
+        evaluation.message("Definition", "sym", symbol, 1)
+        return
+    attributes = evaluation.definitions.get_attributes(name)
+    all = evaluation.definitions.get_definition(name)
+    if attributes:
+        attributes_list = attributes_bitset_to_list(attributes)
+        lines.append(
+            Expression(
+                SymbolHoldForm,
+                Expression(
+                    SymbolSet,
+                    Expression(SymbolAttributes, symbol),
+                    to_mathics_list(*attributes_list, elements_conversion_fn=Symbol),
+                ),
+            )
+        )
+
+    if not A_READ_PROTECTED & attributes:
+        try:
+            print_rules(evaluation.definitions.get_user_definition(name, create=False))
```

Comment:
Similarly `print_rules` is a bad name because nothing is being printed. Instead, we are accumulating in `lines` a list of `Expressions` for the user definition.

Reply:
Now this function is called `gather_rules`. Again, if a better name comes up later, we can change it.